### ISSUE TITLE
implementing set next window constaint

### ISSUE THREAD

**Issue by CodeDoes** [swistakm/pyimgui#93]
I want to use this feature is there a reason it was never added?

**Comment by lubieowoce** [comment]
Most likely it's just that no one got around to it! Let me take a look.

`void          SetNextWindowSizeConstraints(const ImVec2& size_min, const ImVec2& size_max, ImGuiSizeCallback custom_callback = NULL, void* custom_callback_data = NULL)`

Looking at the signature, it's a mixed bag. If you only want to use `size_min`/`size_max`, the wrapper is trivial. But getting the callback to work will probably be harder, since you have to turn a Python callback into a C++ callback. (it should be doable, but might be a bit tricky)

If `size_min`/`size_max` are enough for you, I can submit a patch today! I'll look into handling the callback, but that might take longer.

**Comment by lubieowoce** [comment]
WIP here: https://github.com/lubieowoce/pyimgui/commits/window-constraints/imgui

I've added a basic version of `set_next_window_constraints(...)` (defined in the `core.pyx` file). It only supports `min/max_width` and `min/max_height` for now. As I expected, the callback stuff is shaping up to be pretty hairy – it works most of the time, but the interactions of GC and non-GC code (calling a Python function from C) cause some nasty edge cases :(

**Comment by lubieowoce** [comment]
I [added](https://github.com/lubieowoce/pyimgui/commit/62bd989bf3621df125f68d0e07423668a68bf9c9) initial support for callbacks, but it's pretty brittle – passing a lambda may result in a segault because (i think) the function object gets garbage collected too early. I explain it in more detail in the code and the commit message.

[This gist](https://gist.github.com/lubieowoce/3e215bac79d6b8dfcda8903633d9e7dd) has a test case that exhibits the problem. (at least on my machine – Windows, CPython 3.5.4. Since it's related to memory management, it may behave differently on other versions/OSes, but the underlying issue is there.).

**Comment by swistakm** [comment]
After digging more into I'm pretty sure this cannot be done easily without a lot of effort or some kind of dedicated support implemented in core library. It also cannot be fully emulated in pure Python because that would require from us knowing position of window before it is even defined.

@lubieowoce, I believe that whatever we do it would be very brittle, complex and hard to understand. I think that right now it is better to leave is where it is (with optional "unsafe" implementation as proposed by you) and additional note in documentation explaining what kind of usage is considered unsafe: any functions that may not exist during next `begin()` so mostly lambdas and anonymously decorated functions (e.g partials).

I would also put all future "strictly unsafe" methods to separate module named `unsafe`. That way we can organize them easily and also keep from the eyes of less experienced users until they are sure that need particular implementation.

**Comment by swistakm** [comment]
By the way: that's a very good catch. I'm pretty sure that I would forget about segfault possibility on lambdas.

**Comment by lubieowoce** [comment]
@swistakm yeah, i tried increasing the refcount and it failed exactly for that reason you describe...
i have two (and a half) ideas to make it safe:

1. let's just maintain a reference to the callback for the rest  of the frame. create a global variable (or an `_IO` attribute?) `_current_frame_callbacks = set()`. when the user calls `set_next_window_size_constraints(..., callback=f)`, we add f to `_current_frame_callbacks`. This way, we can keep the function object alive for as long as we want. later, at the end of the frame, we'd clear `_current_frame_callbacks`. I'm not crazy about fixing state problems by adding more state, but it should work.

2. all our problems come from the separation between `unsafe_set_next_window_size_constraints(..., callback=f)` and `begin(...)`. the easiest way to get around it would be to create

```python
def begin_with_size_constraints(..., callback):
     unsafe_set_next_window_size_constraints(..., callback=callback)
    begin(...)
```
it's not the nicest API, but would keep the user within the memory-safe realm.

if we offered context managers for window creation (to abstract away `begin/end` calls¹), it could be made nicer and more composable – something like
```
with window('hello', ...).size_constraints(..., callback=cb):
   draw_stuff()
```
a `window` would only call `begin()` during `__enter__`, and we'd get the `size_constraints` earlier, so we could arrange the calls appropriately.

---

> By the way: that's a very good catch. I'm pretty sure that I would forget about segfault possibility on lambdas.

i write Haskell as well, so i love my lambdas and partials ;)

---
¹ i think I emailed you about this like a year ago, and then forgot completely!

**Comment by lubieowoce** [comment]
@swistakm if we go with `unsafe`, what should the module structure look like? `unsafe` and `core` need to share some type definitions and utility functions, so I was thinking something like this:

```
  core (safe, wraps unsafe)
   | \
   | unsafe
   |   |
  common (types, utils)
```

**Comment by lubieowoce** [comment]
@swistakm I've had some problems with getting multiple Cython modules in one package to compile. I'd propose an easier, if slightly messier solution:

```
# implementation module
[   core.pyx   ]
  cdef safe_function_1(...): ...
  cdef safe_function_2(...):
      if ok(x):
          return unsafe_function(x)
  cdef unsafe_function(...): ...

# namespace modules

[   unsafe.py   ]
  from core import unsafe_function

[   __init__.py   ]
  from core import (
      safe_function_1,
      safe_function_2,
  )
  import unsafe
```

This way, we can use unsafe functions within `core` to provide safe wrappers without futzing around with module dependencies - we just don't expose them in the main `imgui` namespace, but through a dedicated `imgui.unsafe` namespace.

**Comment by swistakm** [comment]
Ok I did experiment with various options and gave it a very long though. We'll use all of your suggestions and probably introduce this gradually:
1. safe/unsafe module split
2. dedicated window context manager with safe size callback guard
3. optional safe set_window_constraint implementation with additional global reference tracking

We currently don't have any other widget accepting callbacks so all of these changes can be done in backwards-compatible way. So I will mark it as 1.1.0 milestone because I want to finally have the 1.0.0 release shipped this week.

**Comment by lubieowoce** [comment]
sorry, i've been away for a while. your way of handling it sounds good! also, let me know if you have anything blocking 1.0.0 that i can help with!
